### PR TITLE
BUG: DataFrame bugs in ndarray assignments

### LINKED ISSUE #51925: BUG: Series in a DataFrame can overwrite other Series in copy action
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
import sys
import pandas as pd
import numpy as np
from numpy import dtype

print('python version: ', sys.version)
print('pandas version: ', pd.__version__)

data_dtype = [('name', dtype(object)), ('value', 'f8'), ('value2', dtype(object)),]
table = pd.DataFrame(np.zeros(0, dtype=data_dtype), index=pd.Index([], dtype=np.int64))
dtypes = table.dtypes
index = np.arange(10, dtype=np.int32)

data_dict = {'value2': np.ones(10, dtype=float), 'value': np.zeros(10, dtype=float),
             'name': np.array([np.arange(13) + x for x in range(10)], dtype=float)}

dd = pd.DataFrame(index=index, columns=table.columns)
dd = dd.assign(**data_dict)
print('**** Original ****')
print(dd.head())
print('**** Bad Copy ****')
print(dd.copy().head())

print('**** 1D ndarray ****')
data_dict['name'] = data_dict['name'][:, 0]
dd = pd.DataFrame(index=index, columns=table.columns)
dd = dd.assign(**data_dict)
print(dd.copy().head())

'''
Program output:
python version:  3.8.16 | packaged by conda-forge | (default, Feb  1 2023, 15:53:35) [MSC v.1929 64 bit (AMD64)]
pandas version:  0+untagged.1.g6169cba
**** Original ****
   name  value  value2
0   0.0    0.0     1.0
1   1.0    0.0     1.0
2   2.0    0.0     1.0
3   3.0    0.0     1.0
4   4.0    0.0     1.0
**** Bad Copy ****
   name  value  value2
0   0.0    2.0     1.0
1   1.0    3.0     2.0
2   2.0    4.0     3.0
3   3.0    5.0     4.0
4   4.0    6.0     5.0
**** 1D ndarray ****
   name  value  value2
0   0.0    0.0     1.0
1   1.0    0.0     1.0
2   2.0    0.0     1.0
3   3.0    0.0     1.0
4   4.0    0.0     1.0
'''
```

### Issue Description

The `name` Series was assigned using a 2D ndarray, and it works fine. But a `DataFrame.copy` makes `name` overwrite `value` and `value2` in the new df. An 1D ndarray will not break the normal behavior.

### Expected Behavior

```text
**** Original ****
   name  value  value2
0   0.0    0.0     1.0
1   1.0    0.0     1.0
2   2.0    0.0     1.0
3   3.0    0.0     1.0
4   4.0    0.0     1.0
**** Bad Copy ****
   name  value  value2
0   0.0    0.0     1.0
1   1.0    0.0     1.0
2   2.0    0.0     1.0
3   3.0    0.0     1.0
4   4.0    0.0     1.0
**** 1D ndarray ****
   name  value  value2
0   0.0    0.0     1.0
1   1.0    0.0     1.0
2   2.0    0.0     1.0
3   3.0    0.0     1.0
4   4.0    0.0     1.0
```

### Installed Versions

Fresh install of development environment following pandas doc gave an error from numba, here's another report:

<details>

INSTALLED VERSIONS
------------------
commit           : 2e218d10984e9919f0296931d92ea851c6a6faf5
python           : 3.8.16.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.22621
machine          : AMD64
processor        : Intel64 Family 6 Model 158 Stepping 10, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : Chinese (Simplified)_China.936
pandas           : 1.5.3
numpy            : 1.22.0
pytz             : 2022.7
dateutil         : 2.8.2
setuptools       : 65.6.3
pip              : 22.3.1
Cython           : None
pytest           : 7.2.2
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : 3.0.9
lxml.etree       : None
html5lib         : None
pymysql          : 1.0.2
psycopg2         : 2.9.5
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : 1.3.5
brotli           : 
fastparquet      : None
fsspec           : 2023.1.0
gcsfs            : None
matplotlib       : 3.6.2
numba            : 0.56.4
numexpr          : 2.8.4
odfpy            : None
openpyxl         : 3.0.10
pandas_gbq       : None
pyarrow          : None
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : 1.10.0
snappy           : None
sqlalchemy       : None
tables           : 3.7.0
tabulate         : None
xarray           : 2023.1.0
xlrd             : None
xlwt             : None
zstandard        : None
tzdata           : None

</details>

### PR DESCRIPTION
**Pull request by determ1ne**
- [x] closes #51925 and #53366
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/v2.1.0.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by determ1ne** [comment]
Any updates on this?

**Comment by determ1ne** [comment]
@mroeschke

File: pandas/core/frame.py [hunk 1]
```diff
@@ -3966,6 +3966,24 @@ def __setitem__(self, key, value):
         ):
             # Column to set is duplicated
             self._setitem_array([key], value)
+        elif (
+            isinstance(value, np.ndarray) and value.ndim > 1 and self.columns.is_unique
+        ):
+            # TODO: a check for MultiIndex should be added
```
Comment (jbrockmendel) [inline pandas/core/frame.py:3972]: does the check on the next line not take care of this TODO?

File: pandas/core/frame.py [hunk 2]
```diff
@@ -3966,6 +3966,24 @@ def __setitem__(self, key, value):
...
+            if isinstance(self.columns, MultiIndex):
+                self._set_item(key, value)
+                return
+            # squeeze 2d ndarray to 1d if possible
+            # this keeps the backward compatibility
+            if np.ndim(value) == 2 and (1 in np.shape(value)):
```
Comment (jbrockmendel) [inline pandas/core/frame.py:3978]: can you use value.ndim and values.shape instead of the numpy methods

File: pandas/core/frame.py [hunk 3]
```diff
@@ -3966,6 +3966,24 @@ def __setitem__(self, key, value):
         ):
             # Column to set is duplicated
             self._setitem_array([key], value)
+        elif (
+            isinstance(value, np.ndarray) and value.ndim > 1 and self.columns.is_unique
```
Comment (jbrockmendel) [inline pandas/core/frame.py:3970]: does it matter what 'key' is here?

File: pandas/tests/frame/test_constructors.py [hunk 4]
```diff
@@ -283,6 +283,11 @@ def test_constructor_cast_failure(self):
         with pytest.raises(ValueError, match=msg):
             df["test"] = np.ones((4, 2))
 
+        # this is not ok
+        msg = "Expected a 1D array, got an array with shape \\(4, 2, 3\\)"
```
Comment (jbrockmendel) [inline pandas/tests/frame/test_constructors.py:287]: this file is for constructor tests.  can you put a dedicated test in tests/frame/indexing/test_setitem.py

File: pandas/core/internals/managers.py [hunk 5]
```diff
@@ -1401,7 +1401,7 @@ def insert(self, loc: int, item: Hashable, value: ArrayLike, refs=None) -> None:
         # insert to the axis; this could possibly raise a TypeError
         new_axis = self.items.insert(loc, item)
 
-        if value.ndim == 2:
+        if value.ndim >= 2:
```
Comment (jbrockmendel) [inline pandas/core/internals/managers.py:1404]: why is this needed?

Comment (determ1ne) [reply pandas/core/internals/managers.py:1404]: > why is this needed?

A ndarray where dim >= 2 will bypass the check here, but still raise an Exception when accessing the DataFrame. For an example for this bug, see #53366

File: pandas/core/frame.py (hunk 1 above)
Comment (determ1ne) [reply pandas/core/frame.py:3972]: > does the check on the next line not take care of this TODO?

No, the check here is just placeholder for further improvements, and the real check should be inserted between "isinstance" line and "_set_item" line. I did not figure out what and how checks for MultiIndex should be added so I skipped this. But I think the current PR is a mitigation covering and fixing frequently used scenes.

**Comment by mroeschke** [comment]
Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.
